apm-lambda-extension/extension/apm_server.go

```diff
@@ -21,35 +21,53 @@ import (
 	"bytes"
 	"compress/gzip"
 	"fmt"
+	"io"
 	"io/ioutil"
 	"log"
 	"net/http"
 )
 
 // todo: can this be a streaming or streaming style call that keeps the
 //       connection open across invocations?
-func PostToApmServer(postBody []byte, config *extensionConfig) error {
+func PostToApmServer(agentData AgentData, config *extensionConfig) error {
 	endpointUri := "intake/v2/events"
-	var compressedBytes bytes.Buffer
-	w := gzip.NewWriter(&compressedBytes)
-	w.Write(postBody)
-	w.Write([]byte{10})
-	w.Close()
+	var req *http.Request
+	var err error
 
-	client := &http.Client{}
+	if agentData.ContentEncoding == "" {
+		pr, pw := io.Pipe()
+		gw, _ := gzip.NewWriterLevel(pw, gzip.BestSpeed)
 
-	req, err := http.NewRequest("POST", config.apmServerUrl+endpointUri, bytes.NewReader(compressedBytes.Bytes()))
-	if err != nil {
-		return fmt.Errorf("failed to create a new request when posting to APM server: %v", err)
+		go func() {
+			_, err = io.Copy(gw, bytes.NewReader(agentData.Data))
+			gw.Close()
+			pw.Close()
+			if err != nil {
+				log.Printf("Failed to compress data: %v", err)
```

Comment:
[minor suggestion]
IIUC, this function will only be executed once the request has been established and is then streamed (via chunked-encoding) to the server.
If that's correct, the message may be a bit misleading as the more common source of failure is that there's a network issue when streaming the data to APM Server.

```suggestion
				log.Printf("Failed to send compressed data: %v", err)
```

Reply:
We are actually not streaming data yet, but rather sending in batches. At this point in the code, the actual failure is that the data could not be compressed in a go routine. If compression failed in that go routine, we will also get an error that is logged on line 52 (`failed to create a new request when posting to APM server`) and the function will return. So I think the error message here is accurate and should stay as-is.

Reply:
I think this streams the batches. See also https://medium.com/stupid-gopher-tricks/streaming-data-in-go-without-buffering-3285ddd2a1e5#2342

Reply:
Yes, you're right that the actual send could be chunked by the go internal transport code, I thought you were referring to the way that we send the data as it's received from the agent. But at this point in the code, I'm calling `io.Copy`, which ends up calling [Write](https://pkg.go.dev/compress/gzip#Writer.Write) on the `gzipWriter` before it gets to the network code. So any compression failures would be returned by `io.Copy`. Network errors would be returned by [resp, err := client.Do(req)](https://github.com/elastic/apm-aws-lambda/pull/72/files#diff-6c889602ebc4f41bed697c18ef699798a40df79b104d827330a8bc38be305324R71).

But that actually brings up another point: the error returned from `client.Do(req)` actually wraps the compression error, which means we don't really need to do anything with the error returned from `io.Copy`. In other words, we could removing the logging of the error and just rely on the error returned in [this line](https://github.com/elastic/apm-aws-lambda/pull/72/files#diff-6c889602ebc4f41bed697c18ef699798a40df79b104d827330a8bc38be305324R73)

Reply:
> I'm calling io.Copy, which ends up calling Write on the gzipWriter before it gets to the network code.

What I'm understanding from the blog that I linked is that `io.Copy` blocks until the `PipeReader` is being read from. This happens when `client.Do(req)` is invoked and the data from the `PipeReader` is written into the HTTP request body via chunked encoding. It's because `io.Copy` blocks that we have to execute it in a concurrently running goroutine. If we didn't do that, we'd never get to the `client.Do(req)` part.

See also the docs for [`io.Pipe`](https://pkg.go.dev/io#Pipe)
> each Write to the PipeWriter blocks until it has satisfied one or more Reads from the PipeReader that fully consume the written data. The data is copied directly from the Write to the corresponding Read (or Reads); there is no internal buffering.